File: project_party/cpp/src/party_api.cpp
// ProjectParty C++ API client implementation (libcurl).
//
// See party_api.hpp for the contract. All methods issue real HTTP calls via a
// libcurl easy handle. Fail-closed: any non-2xx response throws
// `PartyApiException` carrying the backend's `error`/`message` text.

#include "party_api.hpp"

#include <curl/curl.h>

#include <cctype>
#include <iomanip>
#include <sstream>

namespace tigerparty {

namespace {

class CurlGlobal {
public:
    CurlGlobal() { curl_global_init(CURL_GLOBAL_DEFAULT); }
    ~CurlGlobal() { curl_global_cleanup(); }
};

CurlGlobal& curlGlobal() {
    static CurlGlobal g;
    return g;
}

struct WriteCtx {
    std::string buf;
};

size_t writeCb(char* ptr, size_t size, size_t nmemb, void* userdata) {
    auto* ctx = static_cast<WriteCtx*>(userdata);
    const size_t total = size * nmemb;
    ctx->buf.append(ptr, total);
    return total;
}

size_t headerCb(char* /*ptr*/, size_t size, size_t nmemb, void* /*userdata*/) {
    return size * nmemb;
}

} // namespace

ProjectPartyApiClient::ProjectPartyApiClient(std::string baseUrl)
    : base_url_(std::move(baseUrl)) {
    curlGlobal();
}

ProjectPartyApiClient::~ProjectPartyApiClient() = default;

std::string ProjectPartyApiClient::encodePathSegment(const std::string& s) {
    std::ostringstream out;
    out << std::hex << std::uppercase << std::setfill('0');
    for (unsigned char c : s) {
        if (std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
            out << static_cast<char>(c);
        } else {
            out << '%' << std::setw(2) << static_cast<int>(c);
        }
    }
    return out.str();
}

std::string ProjectPartyApiClient::encodeQuery(const std::string& s) {
    // Query values use the same percent-encoding as path segments.
    return encodePathSegment(s);
}

nlohmann::json ProjectPartyApiClient::request(const std::string& path,
                                              const std::string& method,
                                              const std::optional<nlohmann::json>& body,
                                              bool authenticated,
                                              bool absolute) {
    CURL* easy = curl_easy_init();
    if (!easy) {
        throw PartyApiException(0, "curl_easy_init failed");
    }
    WriteCtx ctx;
    struct curl_slist* headers = nullptr;
    std::string bodyStr;
    try {
        headers = curl_slist_append(headers, "Accept: application/json");
        if (body.has_value()) {
            bodyStr = body->dump();
            headers = curl_slist_append(headers, "Content-Type: application/json; charset=utf-8");
        }
        if (authenticated) {
            if (!token_.has_value()) {
                curl_slist_free_all(headers);
                curl_easy_cleanup(easy);
                throw PartyApiException(401, "not authenticated: no JWT token set");
            }
            const std::string auth = "Authorization: Bearer " + *token_;
            headers = curl_slist_append(headers, auth.c_str());
        }
        std::string url;
        if (absolute) {
            std::string base = base_url_;
            const std::string suffix = "/api/v1";
            if (base.size() >= suffix.size() &&
                base.compare(base.size() - suffix.size(), suffix.size(), suffix) == 0) {
                base.erase(base.size() - suffix.size());
            }
            url = base + path;
        } else {
            url = base_url_ + path;
        }
        curl_easy_setopt(easy, CURLOPT_URL, url.c_str());
        curl_easy_setopt(easy, CURLOPT_CUSTOMREQUEST, method.c_str());
        curl_easy_setopt(easy, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(easy, CURLOPT_TIMEOUT, 30L);
        curl_easy_setopt(easy, CURLOPT_CONNECTTIMEOUT, 15L);
        curl_easy_setopt(easy, CURLOPT_FOLLOWLOCATION, 0L);
        curl_easy_setopt(easy, CURLOPT_WRITEFUNCTION, writeCb);
        curl_easy_setopt(easy, CURLOPT_WRITEDATA, &ctx);
        curl_easy_setopt(easy, CURLOPT_HEADERFUNCTION, headerCb);
        curl_easy_setopt(easy, CURLOPT_NOSIGNAL, 1L);
        if (body.has_value()) {
            curl_easy_setopt(easy, CURLOPT_POSTFIELDS, bodyStr.c_str());
            curl_easy_setopt(easy, CURLOPT_POSTFIELDSIZE,
                             static_cast<long>(bodyStr.size()));
        }
        CURLcode rc = curl_easy_perform(easy);
        if (rc != CURLE_OK) {
            curl_slist_free_all(headers);
            curl_easy_cleanup(easy);
            throw PartyApiException(0, std::string("curl: ") + curl_easy_strerror(rc));
        }
        long httpCode = 0;
        curl_easy_getinfo(easy, CURLINFO_RESPONSE_CODE, &httpCode);
        curl_slist_free_all(headers);
        curl_easy_cleanup(easy);
        if (httpCode < 200 || httpCode > 299) {
            std::string msg;
            try {
                auto j = nlohmann::json::parse(ctx.buf.empty() ? "{}" : ctx.buf);
                if (j.contains("error") && j["error"].is_string()) {
                    msg = j["error"].get<std::string>();
                } else if (j.contains("message") && j["message"].is_string()) {
                    msg = j["message"].get<std::string>();
                }
            } catch (...) {
                msg = ctx.buf;
            }
            if (msg.empty()) msg = "HTTP " + std::to_string(httpCode);
            throw PartyApiException(static_cast<int>(httpCode), msg);
        }
        if (ctx.buf.empty()) return nlohmann::json{};
        return nlohmann::json::parse(ctx.buf);
    } catch (...) {
        if (headers) curl_slist_free_all(headers);
        curl_easy_cleanup(easy);
        throw;
    }
}

nlohmann::json ProjectPartyApiClient::get(const std::string& path, bool authenticated) {
    return request(path, "GET", std::nullopt, authenticated, false);
}

nlohmann::json ProjectPartyApiClient::del(const std::string& path) {
    return request(path, "DELETE", std::nullopt, true, false);
}

nlohmann::json ProjectPartyApiClient::post(const std::string& path, const nlohmann::json& body) {
    return request(path, "POST", body, true, false);
}

nlohmann::json ProjectPartyApiClient::put(const std::string& path, const nlohmann::json& body) {
    return request(path, "PUT", body, true, false);
}

nlohmann::json ProjectPartyApiClient::postNoBody(const std::string& path) {
    return request(path, "POST", std::nullopt, true, false);
}

// -- Health --

nlohmann::json ProjectPartyApiClient::getHealth() {
    return request("/health", "GET", std::nullopt, false, true);
}

// -- Auth --

PartyAuthResponse ProjectPartyApiClient::registerUser(const std::string& username,
                                                       const std::string& password) {
    auto res = request("/auth/register", "POST",
                       nlohmann::json{{"username", username}, {"password", password}},
                       false, false);
    PartyAuthResponse out{};
    if (res.contains("token") && res["token"].is_string()) {
        out.token = res["token"].get<std::string>();
        token_ = out.token;
    }
    out.username = res.value("username", "");
    out.role = res.value("role", "");
    return out;
}

PartyAuthResponse ProjectPartyApiClient::login(const std::string& username,
                                               const std::string& password) {
    auto res = request("/auth/login", "POST",
                       nlohmann::json{{"username", username}, {"password", password}},
                       false, false);
    PartyAuthResponse out{};
    if (res.contains("token") && res["token"].is_string()) {
        out.token = res["token"].get<std::string>();
        token_ = out.token;
    }
    out.username = res.value("username", "");
    out.role = res.value("role", "");
    return out;
}

// -- Discovery (public) --

nlohmann::json ProjectPartyApiClient::getCoins() { return get("/coins", false); }
nlohmann::json ProjectPartyApiClient::searchTokens(const std::string& q) {
    return get("/search?q=" + encodeQuery(q), false);
}
nlohmann::json ProjectPartyApiClient::getFeatured() { return get("/featured", false); }
nlohmann::json ProjectPartyApiClient::getTrending() { return get("/trending", false); }
nlohmann::json ProjectPartyApiClient::getMarket() { return get("/market", false); }

// -- Tokens --

nlohmann::json ProjectPartyApiClient::listTokens(const std::optional<std::string>& status) {
    return get(status ? "/tokens?status=" + encodeQuery(*status) : "/tokens");
}
nlohmann::json ProjectPartyApiClient::getToken(const std::string& id) {
    return get("/tokens/" + encodePathSegment(id));
}
nlohmann::json ProjectPartyApiClient::createToken(const nlohmann::json& data) {
    return post("/tokens", data);
}
nlohmann::json ProjectPartyApiClient::updateToken(const std::string& id, const nlohmann::json& data) {
    return put("/tokens/" + encodePathSegment(id), data);
}
nlohmann::json ProjectPartyApiClient::deleteToken(const std::string& id) {
    return del("/tokens/" + encodePathSegment(id));
}
nlohmann::json ProjectPartyApiClient::submitToken(const std::string& id) {
    return postNoBody("/tokens/" + encodePathSegment(id) + "/submit");
}
nlohmann::json ProjectPartyApiClient::approveToken(const std::string& id) {
    return postNoBody("/tokens/" + encodePathSegment(id) + "/approve");
}
nlohmann::json ProjectPartyApiClient::rejectToken(const std::string& id) {
    return postNoBody("/tokens/" + encodePathSegment(id) + "/reject");
}

// -- Listings --

nlohmann::json ProjectPartyApiClient::listListings(const std::optional<std::string>& status) {
    return get(status ? "/listings?status=" + encodeQuery(*status) : "/listings");
}
nlohmann::json ProjectPartyApiClient::getListing(const std::string& id) {
    return get("/listings/" + encodePathSegment(id));
}
nlohmann::json ProjectPartyApiClient::createListing(const nlohmann::json& data) {
    return post("/listings", data);
}
nlohmann::json ProjectPartyApiClient::updateListingStatus(const std::string& id,
                                                          const std::string& status) {
    return put("/listings/" + encodePathSegment(id) + "/status",
               nlohmann::json{{"status", status}});
}
nlohmann::json ProjectPartyApiClient::featureListing(const std::string& id) {
    return postNoBody("/listings/" + encodePathSegment(id) + "/featured");
}

// -- Launchpad --

nlohmann::json ProjectPartyApiClient::listLaunchpads(const std::optional<std::string>& status) {
    return get(status ? "/launchpad?status=" + encodeQuery(*status) : "/launchpad");
}
nlohmann::json ProjectPartyApiClient::getLaunchpad(const std::string& id) {
    return get("/launchpad/" + encodePathSegment(id));
}
nlohmann::json ProjectPartyApiClient::createLaunchpad(const nlohmann::json& data) {
    return post("/launchpad/create", data);
}
nlohmann::json ProjectPartyApiClient::contribute(const std::string& id, const std::string& amount) {
    return post("/launchpad/" + encodePathSegment(id) + "/contribute",
                nlohmann::json{{"amount", amount}});
}
nlohmann::json ProjectPartyApiClient::claimTokens(const std::string& id) {
    return postNoBody("/launchpad/" + encodePathSegment(id) + "/claim");
}
nlohmann::json ProjectPartyApiClient::cancelLaunchpad(const std::string& id) {
    return postNoBody("/launchpad/" + encodePathSegment(id) + "/cancel");
}

// -- Market-making --

nlohmann::json ProjectPartyApiClient::getMakerOrders(const std::optional<std::string>& tokenId) {
    return get(tokenId ? "/market-making/orders?token_id=" + encodeQuery(*tokenId)
                       : "/market-making/orders");
}
nlohmann::json ProjectPartyApiClient::getMarketMakerStatus(const std::string& tokenId) {
    return get("/market-making/status/" + encodePathSegment(tokenId));
}
nlohmann::json ProjectPartyApiClient::createMakerOrders(const nlohmann::json& data) {
    return post("/market-making/orders", data);
}
nlohmann::json ProjectPartyApiClient::updateOrderStatus(const std::string& id,
                                                        const std::string& status) {
    return put("/market-making/orders/" + encodePathSegment(id) + "/status",
               nlohmann::json{{"status", status}});
}
nlohmann::json ProjectPartyApiClient::addLiquidity(const nlohmann::json& data) {
    return post("/market-making/liquidity/add", data);
}
nlohmann::json ProjectPartyApiClient::removeLiquidity(const nlohmann::json& data) {
    return post("/market-making/liquidity/remove", data);
}

// -- Pricing --

nlohmann::json ProjectPartyApiClient::getTokenPrice(const std::string& tokenId) {
    return get("/pricing/" + encodePathSegment(tokenId));
}
nlohmann::json ProjectPartyApiClient::getPriceHistory(const std::string& tokenId) {
    return get("/pricing/history/" + encodePathSegment(tokenId));
}
nlohmann::json ProjectPartyApiClient::setTokenPrice(const std::string& tokenId,
                                                    const std::string& price) {
    return post("/pricing/set", nlohmann::json{{"token_id", tokenId}, {"price", price}});
}
nlohmann::json ProjectPartyApiClient::updatePrice(const std::string& tokenId,
                                                  const std::string& price) {
    return post("/pricing/update", nlohmann::json{{"token_id", tokenId}, {"price", price}});
}

// -- Analytics (public) --

nlohmann::json ProjectPartyApiClient::getTradingVolume() { return get("/analytics/volume", false); }
nlohmann::json ProjectPartyApiClient::getLiquidity() { return get("/analytics/liquidity", false); }
nlohmann::json ProjectPartyApiClient::getHolderCount() { return get("/analytics/holders", false); }
nlohmann::json ProjectPartyApiClient::getTransactionCount() { return get("/analytics/transactions", false); }

// -- Compliance --

nlohmann::json ProjectPartyApiClient::getAuditStatus(const std::string& tokenId) {
    return get("/compliance/audit/" + encodePathSegment(tokenId));
}
nlohmann::json ProjectPartyApiClient::getKYCStatus(const std::string& tokenId) {
    return get("/compliance/kyc/" + encodePathSegment(tokenId));
}
nlohmann::json ProjectPartyApiClient::requestAudit(const nlohmann::json& data) {
    return post("/compliance/audit", data);
}
nlohmann::json ProjectPartyApiClient::submitKYC(const nlohmann::json& data) {
    return post("/compliance/kyc/submit", data);
}

// -- Fees --

nlohmann::json ProjectPartyApiClient::getListingFees() { return get("/fees", false); }
nlohmann::json ProjectPartyApiClient::calculateFees(const nlohmann::json& data) {
    return post("/fees/calculate", data);
}
nlohmann::json ProjectPartyApiClient::payFees(const nlohmann::json& data) {
    return post("/fees/pay", data);
}

// -- Favorites (auth) --

nlohmann::json ProjectPartyApiClient::listFavorites() { return get("/favorites"); }
nlohmann::json ProjectPartyApiClient::addFavorite(const nlohmann::json& data) {
    return post("/favorites", data);
}
nlohmann::json ProjectPartyApiClient::removeFavorite(const std::string& id) {
    return del("/favorites/" + encodePathSegment(id));
}

} // namespace tigerparty

File: project_party/rust/src/party_api.rs
//! ProjectParty Rust client.
//!
//! Async reqwest client delegating to the standalone ProjectParty backend
//! (`project_party/go/cmd/main.go`, port 8106, path prefix `/api/v1`, JWT auth
//! + RBAC). The token is held in-process (`Arc<RwLock<Option<String>>>`) — Rust
//! clients are server/CLI and have no keychain dependency by default; callers
//! that need persistence can wrap `set_token`/`token`.
//!
//! Every method issues a real reqwest call against the backend — no stubs,
//! fakes, or mock data. On any non-2xx response the method returns an `Err`
//! carrying the backend's `error` message (fail-closed); it never returns
//! fabricated data.
//!
//! Method set matches `project_party/web/src/services/api.ts` + the discovery,
//! pricing, analytics, compliance routes the task requires.

use reqwest::{Client, Method};
use serde_json::Value;
use std::sync::{Arc, RwLock};
use std::time::Duration;

pub const PARTY_API_DEFAULT_URL: &str = "http://localhost:8106/api/v1";

#[derive(Debug)]
pub enum PartyError {
    Network(String),
    Decode(String),
    Http { status: u16, message: String },
    Unauthorized,
}

impl std::fmt::Display for PartyError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PartyError::Network(m) => write!(f, "network: {m}"),
            PartyError::Decode(m) => write!(f, "decode: {m}"),
            PartyError::Http { status, message } => write!(f, "http {status}: {message}"),
            PartyError::Unauthorized => write!(f, "not authenticated: no JWT token set"),
        }
    }
}

impl std::error::Error for PartyError {}

pub struct PartyClient {
    http: Client,
    base_url: String,
    token: Arc<RwLock<Option<String>>>,
}

impl PartyClient {
    pub fn new(base_url: impl Into<String>, token: Option<String>) -> Self {
        let http = Client::builder()
            .timeout(Duration::from_secs(30))
            .pool_max_idle_per_host(16)
            .tcp_nodelay(true)
            .build()
            .expect("reqwest client build");
        Self {
            http,
            base_url: base_url.into(),
            token: Arc::new(RwLock::new(token)),
        }
    }

    pub fn default() -> Self {
        Self::new(PARTY_API_DEFAULT_URL, None)
    }

    pub fn set_token(&self, token: Option<String>) {
        *self.token.write().unwrap() = token;
    }

    pub fn token(&self) -> Option<String> {
        self.token.read().unwrap().clone()
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    // -- internal request helpers --

    fn url(&self, path: &str, absolute: bool) -> String {
        if absolute {
            let base = self.base_url.trim_end_matches("/api/v1");
            format!("{}{}", base, path)
        } else {
            format!("{}{}", self.base_url, path)
        }
    }

    async fn request(
        &self,
        method: Method,
        path: &str,
        body: Option<Value>,
        authenticated: bool,
        absolute: bool,
    ) -> Result<Value, PartyError> {
        let url = self.url(path, absolute);
        let mut req = self.http.request(method, &url);
        if authenticated {
            if let Some(t) = self.token() {
                req = req.bearer_auth(t);
            } else {
                return Err(PartyError::Unauthorized);
            }
        }
        if let Some(b) = body {
            req = req.json(&b);
        }
        let resp = req.send().await.map_err(|e| PartyError::Network(e.to_string()))?;
        let status = resp.status();
        let bytes = resp
            .bytes()
            .await
            .map_err(|e| PartyError::Network(e.to_string()))?;
        let text = String::from_utf8_lossy(&bytes);
        let value: Value = if text.trim().is_empty() {
            Value::Null
        } else {
            serde_json::from_str(&text).map_err(|e| PartyError::Decode(e.to_string()))?
        };
        if !status.is_success() {
            let msg = value
                .get("error")
                .or_else(|| value.get("message"))
                .and_then(|v| v.as_str())
                .map(|s| s.to_string())
                .unwrap_or_else(|| {
                    if text.trim().is_empty() {
                        format!("HTTP {}", status.as_u16())
                    } else {
                        text.to_string()
                    }
                });
            return Err(PartyError::Http {
                status: status.as_u16(),
                message: msg,
            });
        }
        Ok(value)
    }

    async fn get(&self, path: &str, authenticated: bool) -> Result<Value, PartyError> {
        self.request(Method::GET, path, None, authenticated, false).await
    }

    async fn get_absolute(&self, path: &str) -> Result<Value, PartyError> {
        self.request(Method::GET, path, None, false, true).await
    }

    async fn send_body(
        &self,
        method: Method,
        path: &str,
        body: Value,
    ) -> Result<Value, PartyError> {
        self.request(method, path, Some(body), true, false).await
    }

    async fn post(&self, path: &str, body: Value) -> Result<Value, PartyError> {
        self.send_body(Method::POST, path, body).await
    }

    async fn put(&self, path: &str, body: Value) -> Result<Value, PartyError> {
        self.send_body(Method::PUT, path, body).await
    }

    async fn delete(&self, path: &str) -> Result<Value, PartyError> {
        self.request(Method::DELETE, path, None, true, false).await
    }

    async fn post_no_body(&self, path: &str) -> Result<Value, PartyError> {
        self.request(Method::POST, path, None, true, false).await
    }

    // -- Health --

    pub async fn get_health(&self) -> Result<Value, PartyError> {
        self.get_absolute("/health").await
    }

    // -- Auth --

    pub async fn register(&self, username: &str, password: &str) -> Result<Value, PartyError> {
        let payload = serde_json::json!({ "username": username, "password": password });
        let res = self
            .request(Method::POST, "/auth/register", Some(payload), false, false)
            .await?;
        if let Some(t) = res.get("token").and_then(|v| v.as_str()) {
            self.set_token(Some(t.to_string()));
        }
        Ok(res)
    }

    pub async fn login(&self, username: &str, password: &str) -> Result<Value, PartyError> {
        let payload = serde_json::json!({ "username": username, "password": password });
        let res = self
            .request(Method::POST, "/auth/login", Some(payload), false, false)
            .await?;
        if let Some(t) = res.get("token").and_then(|v| v.as_str()) {
            self.set_token(Some(t.to_string()));
        }
        Ok(res)
    }

    // -- Discovery (public) --

    pub async fn get_coins(&self) -> Result<Value, PartyError> {
        self.get("/coins", false).await
    }

    pub async fn search_tokens(&self, q: &str) -> Result<Value, PartyError> {
        self.get(&format!("/search?q={}", encode(q)), false).await
    }

    pub async fn get_featured(&self) -> Result<Value, PartyError> {
        self.get("/featured", false).await
    }

    pub async fn get_trending(&self) -> Result<Value, PartyError> {
        self.get("/trending", false).await
    }

    pub async fn get_market(&self) -> Result<Value, PartyError> {
        self.get("/market", false).await
    }

    // -- Tokens --

    pub async fn list_tokens(&self, status: Option<&str>) -> Result<Value, PartyError> {
        let path = match status {
            Some(s) => format!("/tokens?status={}", encode(s)),
            None => "/tokens".to_string(),
        };
        self.get(&path, true).await
    }

    pub async fn get_token(&self, id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/tokens/{}", encode(id)), true).await
    }

    pub async fn create_token(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/tokens", data).await
    }

    pub async fn update_token(&self, id: &str, data: Value) -> Result<Value, PartyError> {
        self.put(&format!("/tokens/{}", encode(id)), data).await
    }

    pub async fn delete_token(&self, id: &str) -> Result<Value, PartyError> {
        self.delete(&format!("/tokens/{}", encode(id))).await
    }

    pub async fn submit_token(&self, id: &str) -> Result<Value, PartyError> {
        self.post_no_body(&format!("/tokens/{}/submit", encode(id))).await
    }

    pub async fn approve_token(&self, id: &str) -> Result<Value, PartyError> {
        self.post_no_body(&format!("/tokens/{}/approve", encode(id))).await
    }

    pub async fn reject_token(&self, id: &str) -> Result<Value, PartyError> {
        self.post_no_body(&format!("/tokens/{}/reject", encode(id))).await
    }

    // -- Listings --

    pub async fn list_listings(&self, status: Option<&str>) -> Result<Value, PartyError> {
        let path = match status {
            Some(s) => format!("/listings?status={}", encode(s)),
            None => "/listings".to_string(),
        };
        self.get(&path, true).await
    }

    pub async fn get_listing(&self, id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/listings/{}", encode(id)), true).await
    }

    pub async fn create_listing(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/listings", data).await
    }

    pub async fn update_listing_status(&self, id: &str, status: &str) -> Result<Value, PartyError> {
        self.put(
            &format!("/listings/{}/status", encode(id)),
            serde_json::json!({ "status": status }),
        )
        .await
    }

    pub async fn feature_listing(&self, id: &str) -> Result<Value, PartyError> {
        self.post_no_body(&format!("/listings/{}/featured", encode(id))).await
    }

    // -- Launchpad --

    pub async fn list_launchpads(&self, status: Option<&str>) -> Result<Value, PartyError> {
        let path = match status {
            Some(s) => format!("/launchpad?status={}", encode(s)),
            None => "/launchpad".to_string(),
        };
        self.get(&path, true).await
    }

    pub async fn get_launchpad(&self, id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/launchpad/{}", encode(id)), true).await
    }

    pub async fn create_launchpad(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/launchpad/create", data).await
    }

    pub async fn contribute(&self, id: &str, amount: &str) -> Result<Value, PartyError> {
        self.post(
            &format!("/launchpad/{}/contribute", encode(id)),
            serde_json::json!({ "amount": amount }),
        )
        .await
    }

    pub async fn claim_tokens(&self, id: &str) -> Result<Value, PartyError> {
        self.post_no_body(&format!("/launchpad/{}/claim", encode(id))).await
    }

    pub async fn cancel_launchpad(&self, id: &str) -> Result<Value, PartyError> {
        self.post_no_body(&format!("/launchpad/{}/cancel", encode(id))).await
    }

    // -- Market-making --

    pub async fn get_maker_orders(&self, token_id: Option<&str>) -> Result<Value, PartyError> {
        let path = match token_id {
            Some(t) => format!("/market-making/orders?token_id={}", encode(t)),
            None => "/market-making/orders".to_string(),
        };
        self.get(&path, true).await
    }

    pub async fn get_market_maker_status(&self, token_id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/market-making/status/{}", encode(token_id)), true)
            .await
    }

    pub async fn create_maker_orders(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/market-making/orders", data).await
    }

    pub async fn update_order_status(&self, id: &str, status: &str) -> Result<Value, PartyError> {
        self.put(
            &format!("/market-making/orders/{}/status", encode(id)),
            serde_json::json!({ "status": status }),
        )
        .await
    }

    pub async fn add_liquidity(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/market-making/liquidity/add", data).await
    }

    pub async fn remove_liquidity(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/market-making/liquidity/remove", data).await
    }

    // -- Pricing --

    pub async fn get_token_price(&self, token_id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/pricing/{}", encode(token_id)), true).await
    }

    pub async fn get_price_history(&self, token_id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/pricing/history/{}", encode(token_id)), true)
            .await
    }

    pub async fn set_token_price(&self, token_id: &str, price: &str) -> Result<Value, PartyError> {
        self.post(
            "/pricing/set",
            serde_json::json!({ "token_id": token_id, "price": price }),
        )
        .await
    }

    pub async fn update_price(&self, token_id: &str, price: &str) -> Result<Value, PartyError> {
        self.post(
            "/pricing/update",
            serde_json::json!({ "token_id": token_id, "price": price }),
        )
        .await
    }

    // -- Analytics (public) --

    pub async fn get_trading_volume(&self) -> Result<Value, PartyError> {
        self.get("/analytics/volume", false).await
    }

    pub async fn get_liquidity(&self) -> Result<Value, PartyError> {
        self.get("/analytics/liquidity", false).await
    }

    pub async fn get_holder_count(&self) -> Result<Value, PartyError> {
        self.get("/analytics/holders", false).await
    }

    pub async fn get_transaction_count(&self) -> Result<Value, PartyError> {
        self.get("/analytics/transactions", false).await
    }

    // -- Compliance --

    pub async fn get_audit_status(&self, token_id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/compliance/audit/{}", encode(token_id)), true)
            .await
    }

    pub async fn get_kyc_status(&self, token_id: &str) -> Result<Value, PartyError> {
        self.get(&format!("/compliance/kyc/{}", encode(token_id)), true)
            .await
    }

    pub async fn request_audit(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/compliance/audit", data).await
    }

    pub async fn submit_kyc(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/compliance/kyc/submit", data).await
    }

    // -- Fees --

    pub async fn get_listing_fees(&self) -> Result<Value, PartyError> {
        self.get("/fees", false).await
    }

    pub async fn calculate_fees(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/fees/calculate", data).await
    }

    pub async fn pay_fees(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/fees/pay", data).await
    }

    // -- Favorites (auth) --

    pub async fn list_favorites(&self) -> Result<Value, PartyError> {
        self.get("/favorites", true).await
    }

    pub async fn add_favorite(&self, data: Value) -> Result<Value, PartyError> {
        self.post("/favorites", data).await
    }

    pub async fn remove_favorite(&self, id: &str) -> Result<Value, PartyError> {
        self.delete(&format!("/favorites/{}", encode(id))).await
    }
}

fn encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.as_bytes() {
        let c = *b as char;
        if c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '.' || c == '~' {
            out.push(c);
        } else {
            out.push_str(&format!("%{:02X}", b));
        }
    }
    out
}
